### PR TITLE
Teach length::Bag how to switch hour cycles

### LINKED ISSUE #671: Hour Cycle Preferences ignored
I can't seem to make the preferences work in the length bag:

```diff
diff --git a/components/datetime/examples/work_log.rs b/components/datetime/examples/work_log.rs
index 390e44b..de434be 100644
--- a/components/datetime/examples/work_log.rs
+++ b/components/datetime/examples/work_log.rs
@@ -11,6 +11,7 @@ icu_benchmark_macros::static_setup!();
 
 use icu_datetime::mock::datetime::MockDateTime;
 use icu_datetime::{options::length, DateTimeFormat};
+use icu_datetime::options::preferences;
 use icu_locid::Locale;
 use icu_locid_macros::langid;
 
@@ -53,6 +54,9 @@ fn main(_argc: isize, _argv: *const *const u8) -> isize {
     let options = length::Bag {
         date: Some(length::Date::Medium),
         time: Some(length::Time::Short),
+        preferences: Some(preferences::Bag {
+            hour_cycle: Some(preferences::HourCycle::H23)
+        }),
         ..Default::default()
     };
```

I'd expect it to make `cargo run --example work_log` display 23 hour time, but it stays 11h.

### PR DESCRIPTION
**Pull request by gregtatum**
Edit: This is now ready for review. Please ignore the first commit as it is #841, which I plan on landing first.

Resolves #671 

Time zones are going to be required for full support here, as the generated patterns do not use the time zones. I commented out a check for unsupported symbols temporarily to make everything work, but the results can be seen in the generated data. What you'll see here is that I opted for generated 2 new sets of patterns, one for h11/h12, and one for h23/h24. This is _in addition_ to the existing styles. the existing `time` style represents the localizer's choice for the best hour cycle. The `time_h11_h12` and `time_h23_24` represent the patterns for those specific preferences. There is a bit of duplicated data doing it this way, but I felt it was more correct as it allows localizers to specify any preference for hour cycle, and it also allows for specific overrides.

There is an assumption that a h11 and h12 symbol can be swapped out correctly in a pattern, as well as a h23 and h24 can be swapped. However, h11 _cannot_ be swapped with with h23 etc.

I do the pattern generation code in the provider to pre-generate the data, rather than requiring the pattern generator to be run at runtime for styles. I anticipate a small regression to memory for this approach, as a trade-off of a bigger one if this pattern manipulation was done at runtime.

### CODE REVIEW

**Comment by zbraniecki** [comment]
Your new sets - `time_h11_h12` and `time_h23_h24` are always a pair of one duplicated and one non-duplicated. Could we skip the duplicated one?

**Comment by gregtatum** [comment]
I took a stab at de-duplicating it. The risk is that lengths mix a `h11h12` with a `h23h24`. In that case I have the provider transform panic. I checked the CLDR data, and no locales are currently doing this. It saves 120 bytes but adds a bit more complexity to the implementation.

| commit | total allocations | max memory |
| ------ | ----------------- | ---------- |
| de-duplicated | 53,718 bytes (+197) | 26,492 bytes (+197)|  |
| duplicated    | 53,838 bytes (+317) | 26,612 bytes (+317) | |
| main branch   | 53,521 bytes        | 26,295 bytes | |

File: components/datetime/src/provider/gregory.rs
Comment (sffc) [inline components/datetime/src/provider/gregory.rs:44]: Suggestion: it would make more sense to me if you had a field that was always h11/h12 and the other always h23/h24, and the preferred_hour_cycle lets you pick between them.  So more like

```rust
    pub time_h12: patterns::LengthPatternsV1,
    pub time_h23: patterns::LengthPatternsV1,
    pub preferred_hour_cycle: pattern::FlexibleHourCycle,
```

File: components/datetime/src/pattern/mod.rs
Comment (sffc) [inline components/datetime/src/pattern/mod.rs:299]: Nit: Hmm, I'm not sure I like the name `FlexibleHourCycle`.  It sounds a bit too close to "flexible day period", which is a whole other beast.  Bikeshed:

- `PatternHourCycle` or `HourCycleForPattern`
- `CoarseHourCycle`
- `FormatHourCycle`

Note that ECMA-402 uses the language "pattern" (h23/h24) and "pattern12" (h11/h12) to distinguish.

https://tc39.es/ecma402/#sec-intl.datetimeformat-internal-slots

File: components/datetime/src/pattern/mod.rs
Comment (sffc) [inline components/datetime/src/pattern/mod.rs:313]: Nit: This file is getting too big.  Please refactor this into its own file.  As a rule of thumb, I like to keep *lib.rs* and *mod.rs* empty except for imports.

File: components/datetime/src/pattern/mod.rs
Comment (sffc) [inline components/datetime/src/pattern/mod.rs:366]: Issue: I think this method is only used in the transformer.  I'd like to explore whether we can get code used only for the transformer out of the core library, or at least behind a feature flag.  I'd be interested in your thoughts on this.

File: components/datetime/src/skeleton.rs
Comment (sffc) [inline components/datetime/src/skeleton.rs:207]: Question: Hmm. Can you help me find where this impl is being used? I'm a bit concerned about exposing an API that may or may not be best practice.

**Review by sffc — CHANGES_REQUESTED** [review]

File: components/datetime/src/provider/gregory.rs
Comment (gregtatum) [reply components/datetime/src/provider/gregory.rs:44]: Ah, that is much nicer.

File: components/datetime/src/pattern/mod.rs
Comment (gregtatum) [reply components/datetime/src/pattern/mod.rs:299]: I like `CoarseHourCycle`.

File: components/datetime/src/skeleton.rs
Comment (gregtatum) [reply components/datetime/src/skeleton.rs:207]: This is being used in the transform with: `apply_coarse_hour_cycle`.

`let skeleton = skeleton::Skeleton::from(&pattern);`

File: components/datetime/src/pattern/mod.rs
Comment (gregtatum) [reply components/datetime/src/pattern/mod.rs:313]: Done.

File: components/datetime/src/pattern/mod.rs
Comment (gregtatum) [reply components/datetime/src/pattern/mod.rs:366]: Adding a `provider_transform_utils` feature flag seems like a reasonable-enough way to hide this from end-users, but still allow us to use internals. The inter-dependencies seem like they could get complicated quick, so a new feature flag seems reasonable. It's nice being able to leverage all of pattern and skeleton utilities to make this stuff work well.

File: components/datetime/src/skeleton.rs
Comment (gregtatum) [reply components/datetime/src/skeleton.rs:207]: This is now hidden behind `provider_transform_utils`.

File: experimental/calendar/README.md
Comment (gregtatum) [inline experimental/calendar/README.md:1]: This was generated by `cargo make generate-readmes`. I'm leaving it since I had this happen a few times now.

File: components/datetime/src/pattern/mod.rs
Comment (sffc) [inline components/datetime/src/pattern/mod.rs:304]: Nit: The default is what is used when no locale data is available, and it is supposed to be as locale-agnostic as possible. I would therefore choose H23H24 since those patterns are less likely to have day periods in them.

File: components/datetime/Cargo.toml
Comment (sffc) [inline components/datetime/Cargo.toml:60]: Nit 1: Can we think of a better name?  Ideas:

- `provider_transform` (generic)
- `transform` (generic)
- `pattern_transformations` (DateTime-specific)

Nit 2: If you choose a sufficiently generic feature name, this feature may be useful in other crates as well, please document it in */components/icu/src/lib.rs* (and re-generate */components/icu/README.md*).

**Review by sffc — COMMENTED** [review]
This looks good!  I like how you were able to support the hour cycle preferences at runtime with a very small amount of code, front-loading most of the work to the transformer.

Just some nits.

File: components/datetime/src/skeleton.rs
Comment (zbraniecki) [reply components/datetime/src/skeleton.rs:262]: `suggestion (n-b)`: You can make maintenance of that logic a bit easier if you make more sweeping assumptions. Something like:

```rust
for let item in pattern.items() {
    if let PatternItem::Field(field) = item {
        let new_field = match field.symbol {
            FieldSymbol::Month(_) => FieldSymbol::Month(fields::Month::Format),
            FieldSymbol::Weekday(_) => FieldSymbol::Weekday(fields::Weekday::Format),
            FieldSymbol::Hour(hour) => match hour {
                Hour::H11 | Hour::H12 => FieldSymbol::Hour(Hour::H12),
                Hour::H23 | Hour::H24 => FieldSymbol::Hour(Hour::H23),
            }

            FieldSymbol::Year(_) | FieldSymbol::Day(_) | FieldSymbol::Weekday(_) |
                FieldSymbol::Minute | FieldSymbol::Second(_) | FieldSymbol::TimeZone(_) => *field.symbol,
        };
        ...
    }
}
```

Comment (zbraniecki) [reply components/datetime/src/skeleton.rs:262]: `question`: where do you sort?

File: components/datetime/src/provider/helpers.rs
Comment (zbraniecki) [inline components/datetime/src/provider/helpers.rs:175]: `suggestion (n-b)`: The function would read better for me if it was chunked into three general steps:
1) Get `time` from either prefs of preferred_hour_cycle
2) Parse it into a pattern
3) If prefs are set, override the hour field.

I haven't try to write it, and the function is small enough, that if you prefer to keep it as-is, I'm not worried about maintenance, so non-blocking suggestion.

File: provider/cldr/src/transform/dates/patterns.rs
Comment (zbraniecki) [inline provider/cldr/src/transform/dates/patterns.rs:263]: `nit (n-b)`: I'm trying to minimize the amount of `unwraps` in our code. Can you use `expect` here? It reads a bit better when errors out.

**Review by zbraniecki — APPROVED** [review]
This looks great! My only concern is that I don't think I understand where you sort the skeleton, but I assume it's my misread. The rest are stylistic/perf suggestions and are up at your discretion.

File: components/datetime/src/pattern/mod.rs
Comment (gregtatum) [reply components/datetime/src/pattern/mod.rs:304]: Works for me.

File: components/datetime/Cargo.toml
Comment (gregtatum) [reply components/datetime/Cargo.toml:60]: Nit 1: I was thinking the name was specific towards "internals required by provider transform", and that it would be useful across the project. I'd prefer something a bit more verbose, such as `provider_transform_internals`. The idea behind this is that it's useful to expose something to the provider transformer code, but not to the public API. I'd like to make it scary to use for our end users, and "internals" suggests it shouldn't be touched.

Nit 2: This is already done:

```
//! - `provider_transform_internals`: This is code that is useful for transforming CLDR data, but
//!   may be risky to ship as a public API. For instance in DateTimeFormat code we may want to
//!   manipulate patterns, but not allow end users to do the same.
```

**Comment by gregtatum** [comment]
@zbraniecki Sorting of the fields is done using the SmallVec API:

```
                if let Err(pos) = fields.binary_search(&field) {
                    fields.insert(pos, field)
                }
```

File: components/datetime/src/skeleton.rs
Comment (gregtatum) [reply components/datetime/src/skeleton.rs:262]: > is this intended for runtime

No, it's only used as a provider transform, and it is explicitly marked as so with the feature `provider_transform_internals`, so it's not perf-sensitive. I'm going to pass on the perf suggestions, as I believe the current implementation is good enough, and I would posit simpler for having one step rather than 3.

> You can make maintenance of that logic a bit easier if you make more sweeping assumptions.

I tweaked things a bit to follow your suggestion here. My process for writing things was pretty organic, but this should make the matching a bit cleaner.

> where do you sort?

File: provider/cldr/src/transform/dates/patterns.rs
Comment (gregtatum) [reply provider/cldr/src/transform/dates/patterns.rs:263]: I did these as wall all other new ones I added.

File: components/datetime/src/provider/helpers.rs
Comment (zbraniecki) [inline components/datetime/src/provider/helpers.rs:175]: `nit, n-b`: you can skip this `return`, since you return anyway.

**Review by zbraniecki — DISMISSED** [review]

**Review by zbraniecki — DISMISSED** [review]

**Review by sffc — DISMISSED** [review]

**Comment by gregtatum** [comment]
I did an ugly fix due to merging in #832. The widths for hour cycles were expanding in ways I didn't want them to. The `get_best_available_format_pattern` wasn't really giving me the results I wanted, as it would respect my width choice. I am debating whether it would be better to have `get_best_available_format_pattern` not take `fields: &[Field]` to operate on, but rather a special subset of the fields, defined as `fields: &[SkeletonField]`.

The skeleton fields do not match 1:1 with pattern fields, and I feel like I'm constantly designing around that. I wonder if I shouldn't file a follow-up issue to tackle this.

**Comment by gregtatum** [comment]
Apologies for the force push, but I consolidated my changes after the approval to make it easier to review. The changes are only e7ee7e3.

**Review by sffc — DISMISSED** [review]
Seems good; I defer to Zibi to review the details of the business logic

**Review by zbraniecki — APPROVED** [review]

**Review by zbraniecki — DISMISSED** [review]

**Review by Manishearth — APPROVED** [review]

**Review by zbraniecki — APPROVED** [review]

**Comment by robertbastian** [comment]
> There is an assumption that a h11 and h12 symbol can be swapped out correctly in a pattern, as well as a h23 and h24 can be swapped. However, h11 cannot be swapped with with h23 etc.

This assumption seems to break in CLDR 43
```
thread '<unnamed>' panicked at 'assertion failed: `(left == right)`
  left: `H23H24`,
 right: `H11H12`: A locale contained a mix of coarse hour cycle types', provider/datagen/src/transform/cldr/datetime/patterns.rs:140:17
```

**Comment by sffc** [comment]
Possibly related:

https://github.com/tc39/ecma402/pull/758

**Comment by sffc** [comment]
@robertbastian What is the locale causing the error? What change in CLDR between 42 and 43 is causing the error? That information can help us decide the best approach to fixing it.

**Comment by robertbastian** [comment]
It's `byn` (https://github.com/unicode-org/cldr-json/blob/maint/maint-43/cldr-json/cldr-dates-full/main/byn/ca-gregorian.json#L285-L289) and `ssy` (https://github.com/unicode-org/cldr-json/blob/maint/maint-43/cldr-json/cldr-dates-full/main/ssy/ca-gregorian.json#L285-L289)

**Comment by robertbastian** [comment]
Do we consider this a CLDR bug, or is this mixed assumption something we can drop?

**Pull request merged**
